apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,323 @@
+                type = "string",
+            },
+        },
+        fetch_interval = {type = "integer", minimum = 1, default = 30},
+        prefix = {type = "string", default = "/nacos/v1/"},
+        page_size = {type = "integer", minimum = 1, default = 100},
+        weight = {type = "integer", minimum = 1, default = 100},
+        timeout = {
+            type = "object",
+            properties = {
+                connect = {type = "integer", minimum = 1, default = 2000},
+                send = {type = "integer", minimum = 1, default = 2000},
+                read = {type = "integer", minimum = 1, default = 5000},
+            },
+            default = {
+                connect = 2000,
+                send = 2000,
+                read = 5000,
+            }
+        },
+    },
+    required = {"host"}
+}
+
+
+local _M = {
+    version = 0.1,
+}
+
+
+local function service_info()
+    local host = local_conf.discovery and
+            local_conf.discovery.nacos and local_conf.discovery.nacos.host
+    if not host then
+        log.error("do not set nacos.host")
+        return
+    end
+
+    local username, password
+    -- TODO Add health check to get healthy nodes.
+    local url = host[math_random(#host)]
+    local auth_idx = str_find(url, "#")
+    if auth_idx then
+        local protocol_idx = str_find(url, "://")
+        local protocol = string_sub(url, 1, protocol_idx + 2)
+        local user_and_password = string_sub(url, protocol_idx + 3, auth_idx - 1)
+        local arr = ngx_re.split(user_and_password, ":")
+        if #arr == 2 then
+            username = arr[1]
+            password = arr[2]
+        end
+        local other = string_sub(url, auth_idx + 1)
+        url = protocol .. other
+    end
+    if local_conf.discovery.nacos.prefix then
+        url = url .. local_conf.discovery.nacos.prefix
+    end
+    if str_byte(url, #url) ~= str_byte("/") then
+        url = url .. "/"
+    end
+
+    return url, username, password
+end
+
+
+local function request(request_uri, path, body, method, basic_auth)
+    local url = request_uri .. path
+    log.info("request url:", url)
+    local headers = core.table.new(0, 0)
+    headers['Connection'] = 'Keep-Alive'
+    headers['Accept'] = 'application/json'
+
+    if basic_auth then
+        headers['Authorization'] = basic_auth
+    end
+
+    if body and 'table' == type(body) then
+        local err
+        body, err = core.json.encode(body)
+        if not body then
+            return nil, 'invalid body : ' .. err
+        end
+        -- log.warn(method, url, body)
+        headers['Content-Type'] = 'application/json'
+    end
+
+    local httpc = http.new()
+    local timeout = local_conf.discovery.nacos.timeout
+    local connect_timeout = timeout.connect
+    local send_timeout = timeout.send
+    local read_timeout = timeout.read
+    log.info("connect_timeout:", connect_timeout, ", send_timeout:", send_timeout,
+        ", read_timeout:", read_timeout, ".")
```

Comment:
Bad style, this would be better:
```suggestion
    log.info("connect_timeout:", connect_timeout, ", send_timeout:", send_timeout,
             ", read_timeout:", read_timeout, ".")
```